include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1080 @@
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+#else
+#include <btree.h>
+#include <phmap.h>
+#endif
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <cstdint>
+#include <initializer_list>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include <dataanalysis.h>
+#include <linalg.h>
+
+namespace fiction
+{
+
+/**
+ * An enumeration of cluster linkage methods. The chosen method defines how clusters are merged in the agglomerative
+ * clustering procedure, by, e.g., defining an inter-cluster distance to minimize for the cluster to merge.
+ */
+enum class sidb_cluster_hierarchy_linkage_method
+{
+    /**
+     * Complete linkage takes the maximum distance between nodes in a cluster.
+     */
+    COMPLETE = 0,
+    /**
+     * Single linkage takes the minimum distance between nodes in a cluster.
+     */
+    SINGLE,
+    /**
+     * Unweighted average linkage takes the average distance between nodes in a cluster, disregarding the number of
+     * elements in a node.
+     */
+    UNWEIGHTED_AVERAGE,
+    /**
+     * Weighted average linkage takes the average distance between nodes in a cluster, weighing in the number of
+     * elements in a node.
+     */
+    WEIGHTED_AVERAGE,
+    /**
+     * Known as Ward's method, this type of linkage merges clusters based on a minimum variance measure.
+     */
+    MINIMUM_VARIANCE
+};
+
+struct sidb_binary_cluster_hierarchy_node;
+using sidb_binary_cluster_hierarchy_node_ptr = std::unique_ptr<sidb_binary_cluster_hierarchy_node>;
+/**
+ * The struct used to store a binary cluster hierarchy that may be used to store the result of the hierarchical
+ * clustering returned by ALGLIB functionality.
+ */
+struct sidb_binary_cluster_hierarchy_node
+{
+    /**
+     * The set of SiDB indices contained in the node
+     */
+#ifdef DEBUG_SIDB_CLUSTER_HIERARCHY
+    std::set<uint64_t> c;
+#else
+    phmap::flat_hash_set<uint64_t> c;
+#endif
+    /**
+     * The two children of the node.
+     */
+    std::array<sidb_binary_cluster_hierarchy_node_ptr, 2> sub;
+    /**
+     * Binary cluster hierarchy node constructor.
+     *
+     * @param sidbs The set of SiDBs to contain in this node.
+     * @param children The pair of binary cluster hierarchy node pointers that become the children of this node.
+     */
+#ifdef DEBUG_SIDB_CLUSTER_HIERARCHY
+    sidb_binary_cluster_hierarchy_node(const std::set<uint64_t>& sidbs,
+#else
+    sidb_binary_cluster_hierarchy_node(phmap::flat_hash_set<uint64_t> sidbs,
+#endif
+                                       std::array<sidb_binary_cluster_hierarchy_node_ptr, 2> children) noexcept :
+            c{std::move(sidbs)},
+            sub{std::move(children)}
+    {}
+};
+/**
+ * This function performs the ALGLIB agglomerative clustering algorithm for a given SiDB layout. By default, the cluster
+ * are created by a minimal positional variance heuristic, also known as Ward's method.
```

Comment:
Can we add a link to a webpage if someone is interested in the ``Ward's Method``?

Reply:
Sounds good!